### DISCUSSION TITLE
Allow 1/n snapping in the editor

### DISCUSSION THREAD

**Discussion by mangomizer** [ppy/osu#16536] [Ideas] [answered]
We have several licensed tracks from multiple FAs that have complex rhythms that are not supported by the beatmap editor in both clients, especially lazer. I acknowledge this is a rarity but I believe that if we are making pre-timed .oszs available to all, it is reasonable to also allow mappers to accurately place objects on the editor without having to resort to weird BPM manipulation.

Recently, DeviousPanda ranked an osu! mapset of II-L - Asteroid Field of DECAPLETS (https://osu.ppy.sh/beatmapsets/1666685#osu/3403124) which is mapped entirely in 1/10 (not supported in either client). He was required to map the song at 2x BPM (1/5) and revert back, resulting in a ranked beatmap that has hundreds of unsnapped objects.

Some more examples:
II-L claims that even 1/11 and 1/13 are insufficient for EXPLORER-5 (https://twitter.com/iiljubeat/status/1478690316350005255)
rN - opia requires 1/14 rhythms in bar 16 (https://twitter.com/rn_dtm/status/1478051466954752000)

As songs continue to get more complex, endlessly iterating a slider to allow for all possible snaps would become too messy. I think the best solution forward would be to implement 1/n snaps as an toggleable option (eg. for experienced mappers only). 

Considering the lazer editor is not yet in a state where you can create ranked sets in a convenient manner, this change would be an extremely welcome addition to stable, as mappers could immediately benefit off of it (yes I know it's feature locked)

**Comment by frenzibyte** [comment, marked as answer]
I believe that is tracked under https://github.com/ppy/osu/issues/2316 already.

**Reply by DeviousPanda** [reply]
that was specifically to add snaps like 1/5, 1/7 and 1/9, this proposal is to basically expand that to allow you to input your own snap divisor (hence the 1/n part)

while the change you linked was helpful, it doesnt rectify the issue still of having fa songs that are essentially not mappable under current editor limitations (and as specifically part of this post is talking about adding this to stable)

**Reply by bdach** [reply]
we can support arbitrary denominators in lazer as a natural extension of the aforementioned feature/thread.

also this repository is specifically intended for discussion about lazer features.

**Reply by frenzibyte** [reply]
The general idea of this discussion thread is allowing unusual timing signatures or beat divisors, and that's already tracked in #2316. 

However, I have missed to note that the part about `1/n` should be mentioned there instead to keep track of it as a solution to the issue. But, it has been touched on there already anyways (https://github.com/ppy/osu/issues/2316#issuecomment-381191850).

~~If this is wanted to be added to osu!(stable) then it should have an issue thread open for it in https://github.com/ppy/osu-stable-issues as well, to simplify tracking feature for both clients instead of mixing the two into one discussion.~~

**Reply by mangomizer** [reply]
It's not really a gamebreaking "issue" and it's more of a "it's within scope/makes sense to implement".
Historically there has been notable resistance to adding more complicated snaps, but some recent FA additions actually use these complex snaps so I think it's appropriate to provide support for this?

I'm not sure if it warrants making a thread on stable, but if you think it's appropriate I'll go ahead and do that.

**Reply by bdach** [reply]
@frenzibyte `osu-stable-issues` is not a valid place to point people at for feature requests. the meaning of "feature-locked" is pretty clear cut.

**Reply by frenzibyte** [reply]
couldn't tell the difference between github "issue" threads and actual issue reports anymore, never mind that.
